src/Sulu/Bundle/AdminBundle/DependencyInjection/Compiler/AddPropertyMetadataMapperPass.php

```diff
@@ -0,0 +1,53 @@
+<?php
+
+/*
+ * This file is part of Sulu.
+ *
+ * (c) Sulu GmbH
+ *
+ * This source file is subject to the MIT license that is bundled
+ * with this source code in the file LICENSE.
+ */
+
+namespace Sulu\Bundle\AdminBundle\DependencyInjection\Compiler;
+
+use Sulu\Bundle\AdminBundle\Metadata\SchemaMetadata\SelectionPropertyMetadataMapper;
+use Sulu\Bundle\AdminBundle\Metadata\SchemaMetadata\SingleSelectionPropertyMetadataMapper;
+use Symfony\Component\DependencyInjection\Compiler\CompilerPassInterface;
+use Symfony\Component\DependencyInjection\ContainerBuilder;
+use Symfony\Component\DependencyInjection\Definition;
+
+class AddPropertyMetadataMapperPass implements CompilerPassInterface
+{
+    public function process(ContainerBuilder $container)
+    {
+        $fieldTypeOptionRegistry = $container->get('sulu_admin.field_type_option_registry');
+        $baseFieldTypes = $fieldTypeOptionRegistry->toArray();
+
+        $this->registerPropertyMetadataMappers(
+            $container,
+            SelectionPropertyMetadataMapper::class,
+            \array_keys($baseFieldTypes['selection'] ?? [])
+        );
+
+        $this->registerPropertyMetadataMappers(
+            $container,
+            SingleSelectionPropertyMetadataMapper::class,
+            \array_keys($baseFieldTypes['single_selection'] ?? [])
+        );
+    }
+
+    /**
+     * @param string[] $fieldTypes
+     */
+    private function registerPropertyMetadataMappers(ContainerBuilder $container, string $className, array $fieldTypes)
+    {
+        foreach ($fieldTypes as $fieldType) {
+            $definition = new Definition($className);
```

Comment:
But then you can't just decorate the `sulu_admin.property_data_mapper.custom_entity_selection` service, if you want some specific schema for your custom entity selection.

Reply:
You still can, just not using the decorate services mechanism from symfony. And I think it does not make it a lot easier in that case anyway.

With symfony decorator:
```xml
    <service id="Some\Mapper"
        decorates="Sulu\Mapper"
    >
        <argument type="service" id=".inner"/>
    </service>
```

Without:

```xml
    <service id="Some\Mapper">
        <argument type="service" id="Sulu\Mapper"/>
        <tag name="property_metadata_mapper" type="custom_entity_selection />
    </service>
```

Haven't used all the right names, because I was too lazy to look them up :see_no_evil: But I think you get my idea :slightly_smiling_face:

Reply:
In theory, yes. But I think most developers are used to use symfony decoration (like me ... until a few days ago I never thought about the possibility to manually use the decoration pattern) and will then be confused, why it's not working as they expect

Reply:
Can we at least try to find a different name than the `className` then? As said, this way we make this service autowirable, and I don't think that's a good idea, since I wouldn't consider that a public service.

That being said, I would personally still do it the other way, in order to keep a few unnecessary services from being created.

Reply:
I don't really have such a strong opinion on that, so I'm also up to implement your initial suggestion. But then I'd like to define the service in the `services.xml` and just add the proper tags in the`SuluAdminExtension`, does that sound good to you?

Reply:
Works for me, if that is also OK for you I would go with it :-)